thewalrus/operations.py

```diff
@@ -0,0 +1,120 @@
+# Copyright 2019 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Operations
+==========
+
+**Module name:** :mod:`thewalrus.operations`
+
+.. currentmodule:: thewalrus.operations
+
+Functions to construct the Fock representation of a Gaussian operation
+represented as a Symplectic matrix and complex displacements.
+
+
+Contains some Gaussian operations and auxiliary functions.
+
+Auxiliary functions
+-------------------
+
+.. autosummary::
+    n_two_mode_squeezed_vac
+    choi_expand
+
+Operations
+---------------
+
+.. autosummary::
+    fock_tensor
+"""
+
+from itertools import product
+import numpy as np
+from thewalrus import hafnian_batched
+from thewalrus.symplectic import expand
+from thewalrus.quantum import Amat
+
+
+def n_two_mode_squeezed_vac(n, r=np.arcsinh(1.0)):
+    r"""
+    Returns the symplectic matrix associated with two mode squeezing operations by amount r between modes i and i+n for 0<=i<n
+
+    Args:
+        n (integer): number of modes
+        r (float): squeezing parameter
+    Returns:
+        (array): the symplectic matrix
+    """
+    ch = np.cosh(r) * np.identity(n)
+    sh = np.sinh(r) * np.identity(n)
+    zh = np.zeros([n, n])
+    Snet = np.block([[ch, sh, zh, zh], [sh, ch, zh, zh], [zh, zh, ch, -sh], [zh, zh, -sh, ch]])
+    return Snet
+
+
+def choi_expand(S, r=np.arcsinh(1.0)):
+    r"""
+    Construct the Gaussian-Choi-Jamiolkowski expansion of a symplectic matrix S
+
+    Args:
+        S (array): symplectic matrix
+
+    Returns:
+        (array): expanded symplectic matrix of twice the size of S
+    """
+
+    (n, m) = S.shape
+    assert n == m
+    assert n % 2 == 0
+    nmodes = n // 2
+    return expand(S, list(range(nmodes)), n) @ n_two_mode_squeezed_vac(nmodes, r)
+
+
+def fock_tensor(S, alpha, cutoff, r=np.arcsinh(1.0)):
+    r"""
+    Calculates the Fock representation of a Gaussian unitary represented by symplectic matrix S
+    and displacements alpha up to cutoff
+    Args:
+        S (array): symplectic matrix
+        alpha (array): complex vector of displacements
+        cutoff (int): cutoff in Fock space
+        r (float): squeezing parameter used for the Choi expansion
+    Return:
+        (array): Tensor containing the Fock representation of the Gaussian unitary
+    """
+    S_exp = choi_expand(S, r)
+    cov = S_exp @ S_exp.T
+    A = Amat(cov)
+    n, _ = A.shape
+    N = n // 2
+    B = A[0:N, 0:N].conj()
```

Comment:
How come we take the conjugate of A here?

Reply:
It is equivalent to taking the second half of the A matrix. More importantly it works!